src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1972,14 +2010,20 @@ RebalanceTableShardsBackground(RebalanceOptions *options, Oid shardReplicationMo
 						 "SELECT pg_catalog.replicate_reference_tables(%s)",
 						 quote_literal_cstr(shardTranferModeLabel));
 		BackgroundTask *task = ScheduleBackgroundTask(jobId, GetUserId(), buf.data,
-													  prevJobIdx, prevJobId);
-		prevJobId[prevJobIdx] = task->taskid;
-		prevJobIdx++;
+													  0, dependTasksList);
+		replicateRefTablesTaskId = task->taskid;
 	}
 
 	PlacementUpdateEvent *move = NULL;
-	bool first = true;
-	int prevMoveIndex = prevJobIdx;
+
+	/*
+	 * A shard move must take a dependency on a previous shecheduled move in the same colocation group.
+	 * Two shards moves with the same colocation id cannot run concurrently.
+	 * If there is no previous move scheduled in the same colocation group the move must
+	 * take a dependency on the reference table replication task if exists otherwise no task dependency is required.
+	 * dependHashMap tracks the last move scheduled for the given colocation id.*/
+	HTAB *dependencyHashMap = InitShardMoveDependencyMap();
```

Comment:
I like this approach. I think if we have 2 more of such hashmaps we're pretty much good: First one containing the nodes. Second one containing the shards.

Reply:
Added nodes hashmap. Thought about shard hashmap but realized that moving the same shards will be sequential due to them being in the same colocation group. So we will always have them ordered. So skipping the shard hashmap.